### ISSUE TITLE
Coverage with NYC

### ISSUE THREAD

**Issue by wbhob** [andywer/ava-ts#7]
I deleted the template because I have a simple question: how do I output coverage with ava-ts?

**Comment by andywer** [comment]
Hey @wbhob. Excellent point...

I played around and tried a bunch of stuff, but the bad news is I am currently not able to get working coverage reports either 😕

Some of the code transparently added by TypeScript or NYC just crashs with `TypeError: Promise resolver undefined is not a function`.

So this is an unsolved issue as of now.

**Comment by harangue** [comment]
I got this working! Here's my `.nycrc`
```json
{
  "extension": [
    ".ts",
    ".tsx"
  ],
  "include": ["src/**/*.ts"],
  "reporter": [
    "text",
    "html"
  ],
  "cache": true,
  "sourceMap": false,
  "instrument": false
}
```

Then I run
```
nyc ava-ts tests/ --verbose
```

I haven't had any issues with this method whatsoever. :)

**Comment by andywer** [comment]
Thanks for sharing, @harangue! :)

Runs without crashing for me, too, but I don't get any coverage in my test project... (Source files are matched by `src/**/*.ts` in my project as well)

<img width="510" alt="bildschirmfoto 2018-01-28 um 10 35 02" src="https://user-images.githubusercontent.com/1842462/35480817-f17369dc-0416-11e8-801c-58b187f16b37.png">

**Comment by harangue** [comment]
Yeah @andywer, it seems I spoke too soon. I got it to run test coverage for all of my files once, somehow, but never again after that. I've been looking at [this blog post](http://rundef.com/typescript-code-coverage-istanbul-nyc) to try to uncover what's wrong but I'm afraid I don't understand all of the mechanics of how ava reports to instanbul.

**Comment by harangue** [comment]
Hmm, here's my nyc config now:
```json
{
    "include": [
      "src/**/*.ts"
    ],
    "exclude": [
      "**/*.d.ts"
    ],
    "extension": [
      ".ts"
    ],
    "require": [
      "ts-node/register"
    ],
    "reporter": [
      "html",
      "text"
    ],
    "sourceMap": false,
    "instrument": true,
    "all": true
}
```
This seems to find all of my files (the `require` option is necessary in addition to the `all` key), but does not register my coverage for any except one of the files.

![](https://image.prntscr.com/image/fhw2ypu2Rv2CCCKj8AtTUQ.png)

Two of the difficulties encountered in the aforementioned blog post seem like potential culprits:

> 1. ts-node was generating source maps containing relative links to the original source files. ts-node and nyc were interpreting these relative paths based on a different-level root folder within the folder hierarchy
> 2. the module used by nyc to read source maps wasn't handling absolute paths correctly. (convert-source-map - source map absolute URLS not handled properly)

**Comment by sigmasoldi3r** [comment]
I have similar issues

+ One is that the coverage report is not telling the correct covered lines of code
+ The other is that when including javascript modules, nyc is unable to find it

So ignoring all files that include the javascript module makes it run, but the coverage report is bad.

```
    "ts-loader": "^3.2.0",
    "ts-node": "^7.0.0", 
    "ava": "^0.25.0",
    "ava-ts": "^0.25.0",
    "typescript": "^2.6.2",
```
Node version: `v10.7.0`

**Comment by jack-guy** [comment]
@sigmasoldi3r Ava supports .ts files now (and coverage reports are working for me too).

**Comment by creaux** [comment]
@jack-guy yes but if you are using babel 7 + typescript then you are still tends on ava-ts. I didn't find different solution. Can you be more explicit about support of ts files?

**Comment by dougmolineux** [comment]
When I exclude:
```
"exclude": [
      "**/*.d.ts"
    ],
```
I am seeing this error:
```
TSError: ⨯ Unable to compile TypeScript
src/app.ts (84,29): Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
```

**Comment by andywer** [comment]
@dougmolineux Still hard to tell if this has to do with AVA-TS or just TypeScript... Are you talking about an `exclude` in the `tsconfig.json` or in the `ava` config in the `package.json`?
